### DISCUSSION TITLE
Question regarding packing of boxes

### DISCUSSION THREAD

**Discussion by hrmtacos** [dvdoug/BoxPacker#241] [Q&A] [answered]
First off, thank you for putting the effort into this. It's a really difficult problem to solve. I'm working on a project right now that does shipping 100% by volume, which is easy, but not really _smart_

In any case the docs say:
"Pack largest (by volume) items first"

If I'm reading this right, and item that is say 5x4x1 (20 cubic inches) would "fit" into a box that's 10x2x2 (40 cubic inches) by _volume_ - which doesn't seem right. Is that accurate?  I'm likely just misreading the docs but I want to make sure.

**Comment by dvdoug** [comment, marked as answer]
Hi

Physical limitations are absolutely respected, the documentation there is referring to the amount of 'smarts' the algorithm uses. The boxes are packed (mostly) in size order, there are just limits on the amount of "box tetris" that is used.

Hope that helps?

**Reply by hrmtacos** [reply]
> Physical limitations are absolutely respected
This is great, that's what I was curious about :)

Thanks!

**Comment by hrmtacos** [comment]
Hey one more question @dvdoug - how do you handle items with decimal points?

One of the big challenges is many of the items we'll see are items like pens, which would be something like 6x0.5x0.5 - box packer seems to round this down to 0 and it causes some weird visualizations. 

I was curious if you had any recommendations.  My devs are testing a few options but I was curious if you'd figured this out

THANK YOU again

**Reply by shakethatweight-simon** [reply]
Our measurements are also floats, so we’ve multiplied by 10 to convert  them to integers. E.g. 10.5 cm is now 105mm just make sure you multiply everything by 10 or 100 or whatever multiple you need to make integers, all box and item sizes.

**Reply by dvdoug** [reply]
> "how do you handle items with decimal points?"

Don't 😉. The interfaces are documented as integer mm and a fraction of a mm is too small to worry about. There's a lot of comparison operations involved in the code and keeping everything as an integer avoids all of the complications that arise when comparing floats.

![image](https://user-images.githubusercontent.com/1571110/132958869-548f97cc-698a-4b2a-bf54-8d4bc2a0539d.png)

> seems to round this down to 0

If you're passing floats, then as BoxPacker declares `int` PHP automatically truncates the values. If you put `declare(strict_types=1); ` then you *should* get warnings about that.

Hope that helps

**Reply by hrmtacos** [reply]
Thanks! Your description makes sense.  Thank you so much!
